graphql/client.go

```diff
@@ -239,3 +334,70 @@ func (c *client) createGetRequest(req *Request) (*http.Request, error) {
 
 	return httpReq, nil
 }
+
+func (w *webSocketClient) StartWebSocket(ctx context.Context) (errChan chan error, err error) {
+	w.conn, err = w.Dialer.DialContext(ctx, w.endpoint, w.Header)
+	if err != nil {
+		return nil, err
+	}
+	err = w.sendInit()
+	if err != nil {
+		w.conn.Close()
+		return nil, err
+	}
+	err = w.waitForConnAck()
+	if err != nil {
+		w.conn.Close()
+		return nil, err
+	}
+	go w.listenWebSocket(ctx)
+	return w.errChan, err
+}
+
+func (w *webSocketClient) CloseWebSocket() {
+	if w.conn == nil {
+		return
+	}
+	defer w.conn.Close()
+	err := w.conn.WriteMessage(closeMessage, formatCloseMessage(closeNormalClosure, ""))
+	if err != nil {
+		w.errChan <- err
+	}
+}
+
+func (w *webSocketClient) Subscribe(req *Request, interfaceChan interface{}, forwardDataFunc ForwardDataFunction) (string, error) {
+	if req.Query != "" {
+		if strings.HasPrefix(strings.TrimSpace(req.Query), "query") {
+			return "", errors.New("client does not support queries")
+		}
+		if strings.HasPrefix(strings.TrimSpace(req.Query), "mutation") {
+			return "", errors.New("client does not support mutations")
+		}
+	}
+
+	subscriptionID := uuid.NewString()
+	subscriptionMsg := webSocketSendMessage{
+		Type:    webSocketTypeSubscribe,
+		Payload: req,
+		ID:      subscriptionID,
+	}
+	err := w.sendStructAsJSON(subscriptionMsg)
+	if err != nil {
+		return "", err
+	}
+	w.subscriptions.Create(subscriptionID, interfaceChan, forwardDataFunc)
+	return subscriptionID, nil
+}
+
+func (w *webSocketClient) Unsubscribe(subscriptionID string) error {
+	completeMsg := webSocketSendMessage{
+		Type: webSocketTypeComplete,
+		ID:   subscriptionID,
+	}
+	err := w.sendStructAsJSON(completeMsg)
+	if err != nil {
+		return err
+	}
+	w.subscriptions.Delete(subscriptionID)
```

Comment:
Then conversely here, if we receive more messages for this sub that were already inflight when we sent the complete, we'll drop them. I think probably we want to keep the subscription ID around (maybe clean up the channel and just no-op). Then we can just discard those instead of erroring.

(Or, we could discard without error on unknown subscription ID. I don't like the idea of dropping errors like that but it's hard to know what the caller would do with them; at best they're useful to find a bug in genqlient or in the server. But probably better to just make that unlikely; we can always make the error `Is`-able if it becomes an issue with some servers.)

Reply:
So now I keep the subscription in the map but with an attribute `hasBeenUnsubscribed` that is set to `true` when the client unsubscribes, and that is checked when receiving a websocket message to decide whether or not forwarding the message to the channel. Does it work for you ?

Reply:
That seems great!